### PR TITLE
lib: nrf_cloud: Handle disconnection request from nRF Cloud

### PR DESCRIPTION
**Pull request by glarsennordic**
nRF Cloud library now triggers nrf_cloud_disconnect() if an MQTT message
from nRF Cloud requesting that the device disconnect is received (for
example, when deleting a device from nRF Cloud portal).

Please note that the disconnection event type will be
NRF_CLOUD_DISCONNECT_USER_REQUEST.

IRIS-3728

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 1]
```diff
@@ -1522,3 +1522,33 @@ int nrf_cloud_parse_cell_pos_response(const char *const buf,
...
+	cJSON *discon_request_obj = cJSON_Parse(buf);
+
+	/* Check for nRF Cloud disconnection request MQTT message */
+	if (!json_item_string_exists(discon_request_obj, NRF_CLOUD_JSON_MSG_TYPE_KEY,
+				     NRF_CLOUD_JSON_MSG_TYPE_VAL_DISCONNECT) ||
+	    !json_item_string_exists(discon_request_obj, NRF_CLOUD_JSON_APPID_KEY,
+				     NRF_CLOUD_JSON_APPID_VAL_DEVICE)) {
+		/* Not a disconnection request message */
+		ret = false;
+		goto cleanup;
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1548]: goto not needed since the label is just below this section

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 2]
```diff
@@ -1522,3 +1522,33 @@ int nrf_cloud_parse_cell_pos_response(const char *const buf,
...
+	}
+
+cleanup:
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1551]: label not needed

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 3]
```diff
@@ -522,6 +522,14 @@ static int dc_rx_data_handler(const struct nct_evt *nct_evt)
 		.topic = nct_evt->param.dc->topic,
 	};
 
+	if(nrf_cloud_detect_disconnection_request(nct_evt->param.dc->data.ptr)) {
+		int err = nrf_cloud_disconnect();
+		if (err < 0) {
+			LOG_ERR("Failed to disconnect from nRF Cloud in response to disconnection request, error: %d", err);
+		}
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:529]: It would be good to add a LOG_DBG() here to indicate that the disconnect was requested from the cloud side.

**Review by plskeggs — APPROVED** [review]
Looks great.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 3 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:529]: Consider it done!

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:529]: I had to shorten that debug line, too, by the way, for compliance reasons, in case that alters your standpoint at all. (I doubt that it does)

Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:530]: It happens.  You can split the string onto the next line, e.g.:
```
    LOG_DBG("here is a really long "
                     "string.  It is absolutely huge.");

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c (hunk 1 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1549]: Understood, I included it to maintain structural consistency with the rest of nrf_cloud_codec, but that was probably not necessary. Will be removed!

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 3 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:533]: Done!

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 4]
```diff
@@ -522,6 +522,18 @@ static int dc_rx_data_handler(const struct nct_evt *nct_evt)
...
+	if (nrf_cloud_detect_disconnection_request(nct_evt->param.dc->data.ptr)) {
+		LOG_DBG("nRF Cloud has requested us to disconnect, "
+			"likely due to device deletion");
+		int err = nrf_cloud_disconnect();
+
+		if (err < 0) {
+			LOG_ERR("nRF Cloud disconnection-on-request "
+				"failure, error: %d", err);
+		}
+		return 0;
```
[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:534]: I think a new disconnect reason would be the least invasive approach, I will pursue that.

[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:534]: We ended up deciding not to add a new disconnect reason, but instead allow apps to still receive the raw JSON message even when detected. This has been implemented and pushed

Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:534]: remove the return here so that `nfsm_set_current_state_and_notify` gets called.
i'm not sure if it matters, but it might be best, event-wise, to perform the disconnect after `nfsm_set_current_state_and_notify` is called?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 5]
```diff
@@ -522,6 +522,18 @@ static int dc_rx_data_handler(const struct nct_evt *nct_evt)
 		.topic = nct_evt->param.dc->topic,
 	};
 
+	if (nrf_cloud_detect_disconnection_request(nct_evt->param.dc->data.ptr)) {
+		LOG_DBG("nRF Cloud has requested us to disconnect, "
+			"likely due to device deletion");
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:527]: you could update this to say with certainty that the device has been removed from nrf cloud account since it appears to be the only reason this message is sent.

**Review by jayteemo — CHANGES_REQUESTED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c (hunk 4 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:534]: Sorry, I must have forgotten to add my changes to the most recent commit. Let me try again.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c]: Should be as you describe now

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 6]
```diff
@@ -522,13 +522,26 @@ static int dc_rx_data_handler(const struct nct_evt *nct_evt)
...
 	}
 
 	nfsm_set_current_state_and_notify(nfsm_get_current_state(), &cloud_evt);
 
+	if (discon_req) {
+		LOG_DBG("Device deleted from nRF Cloud");
+		int err = nrf_cloud_disconnect();
+
+		if (err < 0) {
+			LOG_ERR("nRF Cloud disconnection-on-delete "
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:539]: i believe the max accepted line length is 100 now... used to be 80.
this log msg can be made into one line.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c]: Done!

**Review by jayteemo — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 7]
```diff
@@ -1522,3 +1522,32 @@ int nrf_cloud_parse_cell_pos_response(const char *const buf,
...
 }
+
+bool nrf_cloud_detect_disconnection_request(const char *const buf)
+{
+	if (buf == NULL) {
+		return false;
+	}
+
+	/* Fast test to avoid parsing EVERY message with cJSON. */
+	if (strstr(buf, NRF_CLOUD_JSON_APPID_VAL_DEVICE) == NULL ||
```
Comment (simensrostad) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1533]: Maybe add a note here that justifies the use of non-length input based functions such as `strstr` and `cJSON_Parse`. Due to the explicit null termination in https://github.com/nrfconnect/sdk-nrf/blob/53a4486ed22b28f0fcbf16dbcfbeca0cf672e3f2/subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c#L855

**Review by simensrostad — APPROVED** [review]

**Review by jtguggedal — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c [hunk 8]
```diff
@@ -522,13 +522,25 @@ static int dc_rx_data_handler(const struct nct_evt *nct_evt)
...
 	nfsm_set_current_state_and_notify(nfsm_get_current_state(), &cloud_evt);
 
+	if (discon_req) {
+		LOG_DBG("Device deleted from nRF Cloud");
+		int err = nrf_cloud_disconnect();
+
+		if (err < 0) {
+			LOG_ERR("nRF Cloud disconnection-on-delete failure, error: %d", err);
+		}
+		return 0;
```
[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:541]: Done!

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c (hunk 7 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1533]: Actually, it may be simpler (and better) to not assume null termination since arbitrary binary may potentially be passed to this function, good point.

Yes, it's true that the MQTT transport guarantees all output is null-terminated, but it is better not to rely on that.

Good catch, thank you.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1540]: I have updated the code to verify that there is a null byte within the first 200 characters of the passed buffer. This should be sufficient, since the JSON messages we are looking out for are typically about 150 characters. Anything not terminated within 200 characters is either not a null-terminated string, or is otherwise not a DISCON message

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 9]
```diff
@@ -1522,3 +1522,39 @@ int nrf_cloud_parse_cell_pos_response(const char *const buf,
...
+bool nrf_cloud_detect_disconnection_request(const char *const buf)
+{
+	if (buf == NULL) {
+		return false;
+	}
+
+	/* The candidate buffer must be a null-terminated string less than
+	 * a certain length
+	 */
+	if (memchr(buf, '\0', NRF_CLOUD_JSON_MSG_MAX_LEN_DISCONNECT) == NULL) {
```
Comment (plskeggs) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1535]: Couldn't you reduce this to the size of the cloud DISCON message?  200 seems too large given that this is a function that is only for checking for that specific message.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1535]: Ah, yes. The message we are looking for is typically 50 characters, not 150 as I thought. I will set the limit to 75.

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:1535]: Done!

**Review by simensrostad — APPROVED** [review]
